src/event.zig

```diff
@@ -3,5 +3,167 @@
+///     .Resize => |new_size| ...
+/// }
+/// ```
+pub const Event = union(EventType) {
+    Key: KeyEvent,
+    Mouse: MouseEvent,
+    Resize: Size,
+};
+
+pub const EventType = enum {
+    Key,
+    Mouse,
+    Resize,
+};
+
+/// A KeyEvent is an event triggered by a key press or release.
+///
+/// **value**: The key that has been pressed.
+/// **state**: Whether the key event is for key up or key down.
+/// **modifier**: One or several modifier keys that were held when the key was pressed.
+/// **repeated**: Applies to non-raw mode, where keys can be held down to be repeated.
+/// When this field is true, this is a repeat of an earlier KeyEvent. KeyState is always
+/// `Down` when a key is a repeat.
+pub const KeyEvent = struct {
+    value: KeyValue,
+    state: KeyState,
+    modifier: ?Modifier,
+    repeated: bool,
+};
+
+/// The Rune field is used when a key is alphanumerical: a-zA-Z, 0-9, and other character
+/// keys present on the keyboard. Some special characters like Tab ('\t') and space (' ')
+/// will be a Rune value. F1-F12 are found in the Function field. Home, ScrollUp, ScrollDown,
+/// Backspace, Return, and more are found in the Control field.
+pub const KeyValue = union(KeyType) {
+    Rune: Rune,
+    Function: KeyFunction,
+    Control: KeyControl,
+};
+
+pub const KeyType = enum {
+    Rune,
+    Function,
+    Control,
+};
+
+/// A modifier is a key that has been held when another key is pressed or a mouse
+/// event has occurred. This enum is a bitfield, so none, one, or any number of
+/// these fields can be selected in the same enum.
+/// ```zig
+/// if (modifier & Modifier.Alt > 0 or modifier & Modifier.RightAlt > 0) {
+///     // An alt key has been pressed ...
+/// }
+/// ```
+pub const Modifier = enum(u7) {
+    Shift = 0b1,
+    Alt = 0b10,
+    RightAlt = 0b100,
+    Ctrl = 0b1000,
+    RightCtrl = 0b10000,
+    Meta = 0b100000,
+    Function = 0b1000000,
+    _
+};
+
+/// **Down**: The key has just been pressed and has not been released yet.
+/// **Released**: The key was pressed earlier and has just been released.
+pub const KeyState = enum {
+    Down,
+    Released,
+};
+
+/// A function key F1-F12.
+pub const KeyFunction = enum {
+    F1,
+    F2,
+    F3,
+    F4,
+    F5,
+    F6,
+    F7,
+    F8,
+    F9,
+    F10,
+    F11,
+    F12,
+};
+
+pub const KeyControl = enum {
+    Backspace,
+    Delete,
+    Return,
+    Escape,
+    Home,
+    End,
+    PageUp,
+    PageDown,
+    Up,
+    Down,
+    Left,
+    Right,
+};
+
+/// An event triggered by a mouse button or scroll wheel.
+///
+/// **position**: The mouse position in the terminal.
+/// **action**: Whether the event was triggered by a Click, scroll, or other means.
+/// **button**: If the action is a Click or DoubleClick, which mouse button was used.
+pub const MouseEvent = struct {
+    position: Position,
+    action: MouseAction,
+    button: ?MouseButton,
+};
+
+pub const MouseAction = enum {
+    Click,
+    DoubleClick,
```

Comment:
```suggestion
    DoubleClick,
    TripleClick,
```
Even if Windows API will not return a `TripleClick` event on its own, it would be good to compare click event timings and return a `TripleClick` when appropriate. Some other TUI libraries offer `TripleClick` support, and it is also a separate behaviour in many GUIs.

Reply:
I've never heard of applications using a triple click... Sounds like weird UX.

Reply:
One common example is a text editor:

Single click = move cursor
Double click = select word
Triple click = select line

Reply:
OH good point, letsdoit